submissions/views/xqueue.py

```diff
@@ -0,0 +1,296 @@
+                        "submission_time": str(int(external_grader.created_at.timestamp())),
+                        "random_seed": 1
+                    }),
+                    "student_response": external_grader.submission.answer
+                }
+
+                payload = {
+                    'xqueue_header': json.dumps({
+                        'submission_id': external_grader.submission.id,
+                        'submission_key': external_grader.pullkey
+                    }),
+                    'xqueue_body': json.dumps(submission_data),
+                    'xqueue_files': json.dumps(get_files_for_grader(external_grader))
+                }
+
+                return Response(
+                    self.compose_reply(True, content=json.dumps(payload)),
+                    status=status.HTTP_200_OK
+                )
+
+            except ValueError as e:
+                return Response(
+                    self.compose_reply(False, f"Error processing submission: {str(e)}"),
+                    status=status.HTTP_400_BAD_REQUEST
+                )
+
+        return Response(
+            self.compose_reply(False, f"Queue '{queue_name}' is empty"),
+            status=status.HTTP_404_NOT_FOUND
+        )
+
+    @action(detail=False, methods=['post'], url_name='put_result')
+    @transaction.atomic
+    def put_result(self, request):
+        """
+        Endpoint for graders to post their results and update submission scores.
+        """
+        (reply_valid, submission_id, submission_key, score_msg, points_earned) = (
+            self.validate_grader_reply(request.data))
+
+        if not reply_valid:
+            log.error("Invalid reply from pull-grader: request.data: %s",
+                      request.data)
+            return Response(
+                self.compose_reply(False, 'Incorrect reply format'),
+                status=status.HTTP_400_BAD_REQUEST
+            )
+
+        try:
+            external_grader = ExternalGraderDetail.objects.select_for_update(
+                                                                        nowait=True).get(submission__id=submission_id)
+        except ExternalGraderDetail.DoesNotExist:
+            log.error(
+                "Grader submission_id refers to nonexistent entry in Submission DB: "
+                "grader: %s, submission_key: %s, score_msg: %s",
+                submission_id,
+                submission_key,
+                score_msg
+            )
+            return Response(
+                self.compose_reply(False, 'Submission does not exist'),
+                status=status.HTTP_404_NOT_FOUND
+            )
+
+        if not external_grader.pullkey or submission_key != external_grader.pullkey:
+            log.error(f"Invalid pullkey: submission key from xwatcher {submission_key} "
+                      f"and submission key stored {external_grader.pullkey} are different")
```

Comment:
Please log the key they tried to send for this entry, since it might aid debugging, e.g. if they're only different by case, or if they're different by whitespace, etc.